### PR TITLE
Schema v2 + Pilot Sessions + Enhancements

### LINKED ISSUE #222: Store images root cids in separate table
For both the layers and rig images, we should store the root cid of each in a separate table, then only store the path within that root in the layers and rigs tables. This allows for easier updating of the root cid if a one or more images is updated, without needing to update every table row. The query used to generate NFT metadata will need to be updated to join against this new table and build the appropriate IPFS path.

### PR DESCRIPTION
**Pull request by asutula**
**NOTE:** I had originally planned to doing the new pilot sessions work as part of a new effort to build more generalized Tableland dev tooling. Around the same time, some new requirements came up that required big changes to this repo. To minimize risk and make sure I could get all that done, I just decided to incorporate the new pilot flight time table stuff in here.  This PR provides tooling to simply create the pilot sessions table, not anything about the controller or writing to it. (cc @dtbuchholz and @datadanne).

We had a couple new requirements for Rigs: The ability to update Rig layer images and the corresponding Rig renders, plus a new table for tracking pilot flight time. The updating of layer/render imagery was actually more complex and required a refactor to the Rigs tables design. That handled in this PR, but it provided a good opportunity to also clean up some other things I've notices and learned since the original launch. In not particular order:

- No more `rigs` table, and a new `lookups` table that stores lookup information related to imagery stored in IPFS. All data that was in `rigs` can be derived from a rig id and the `lookups` table. The increased normalization of data makes updating rigs imagery easier and cheaper, and overall rigs storage cost on Tableland much cheaper.
- Better/more consistent image names in IPFS
- Better/more consistent naming for image related folders and flags for the CLI
- Additional "medium" size image render (with alpha version as well), so now we have "full" , "medium", and "thumb"
- Cleaned up rendering pipeline code, much easier to understand, less bug prone
- Removed some generated JS code from the repo

Once this is merged, I plan on rebasing my transaction cost tracking brach, then we should newly publish the Rigs data to all networks we want it on using this new schema, transactions logged, etc.

Closes #222 #223 #209 #210

### CODE REVIEW

File: pkg/storage/tableland/store.go [hunk 1]
```diff
@@ -60,17 +48,58 @@ var (
...
+			layers_cid text,
+			image_full_name text,
+			image_full_alpha_name text,
+			image_medium_name text,
+			image_medium_alpha_name text,
+			image_thumb_name text,
+			image_thumb_alpha_name text,
+			animation_base_url text
+		)`,
+	}
```
Comment (asutula) [inline pkg/storage/tableland/store.go:65]: @sanderpick I want to get some quick and early thumbs up or down from you about this change to the rigs tables design. To summarize, above you can see the removal of the `rigs` table, and the creation of a new `lookups` table. This is a more normalized design that optimized more for data compactness, cost savings and ease of updates. I'll leave another comment in the file showing how the metadata query changes accordingly.

File: ethereum/helpers/uris.ts [hunk 2]
```diff
@@ -1,22 +1,23 @@
...
-        `select json_object('name','Rig #'||id,'external_url','https://tableland.xyz/rigs/'||id,'image',image,'image_alpha',image_alpha,'thumb',thumb,'thumb_alpha',thumb_alpha,'attributes',json_group_array(json_object('display_type',display_type,'trait_type',trait_type,'value',value))) from ${tokensTable} join ${attributesTable} on ${tokensTable}.id=${attributesTable}.rig_id where id=ID group by id;`
+        `select json_object('name','Rig #'||rig_id,'external_url','https://tableland.xyz/rigs/'||rig_id,'image','ipfs://'||renders_cid||'/'||rig_id||'/'||image_full_name,'image_alpha','ipfs://'||renders_cid||'/'||rig_id||'/'||image_full_alpha_name,'image_medium','ipfs://'||renders_cid||'/'||rig_id||'/'||image_medium_name,'image_medium_alpha','ipfs://'||renders_cid||'/'||rig_id||'/'||image_medium_alpha_name,'thumb','ipfs://'||renders_cid||'/'||rig_id||'/'||image_thumb_name,'thumb_alpha','ipfs://'||renders_cid||'/'||rig_id||'/'||image_thumb_alpha_name,'animation_url',animation_base_url||rig_id,'attributes',json_group_array(json_object('display_type',display_type,'trait_type',trait_type,'value',value))) from ${attributesTable} join ${lookupsTable} where rig_id=ID group by rig_id;`
```
Comment (asutula) [inline ethereum/helpers/uris.ts:20]: @sanderpick just to show how the metadata query can work with only `rigs_attributes` and `lookups` tables. The query is longer, but that is the price to pay for the benefit of cost savings and to be able to easily and cheaply do something like change the root cid of the layers images or rigs renders.

Comment (sanderpick) [reply ethereum/helpers/uris.ts:20]: is `renders_cid` the root cid?

File: ethereum/deployments.ts [hunk 3]
```diff
@@ -17,8 +17,10 @@ export interface RigsDeployment {
   allowlistTable: string;
   partsTable: string;
   layersTable: string;
-  tokensTable: string;
   attributesTable: string;
+  lookupsTable: string;
+  pilotSessionsTable: string;
+  displayAttributes: boolean;
```
Comment (asutula) [inline ethereum/deployments.ts:23]: This file needs to be updated once we publish the new Rigs schema to the various chains. #238

File: cmd/rigs/cmd/publish_data.go [hunk 4]
```diff
@@ -218,3 +180,35 @@ func attrsJobs(ctx context.Context, s local.Store, jobID *int) ([]wpool.Job, err
...
+				ImageFullAlphaName:   "image_full_alpha.png",
+				ImageMediumName:      "image_medium.png",
+				ImageMediumAlphaName: "image_medium_alpha.png",
+				ImageThumbName:       "image_thumb.png",
+				ImageThumbAlphaName:  "image_thumb_alpha.png",
+				AnimationBaseURL:     "https://tableland.xyz/rigs/animate?id=",
+			}
+			if err := store.InsertLookups(ctx, lookups); err != nil {
+				return nil, fmt.Errorf("calling insert lookups: %v", err)
+			}
```
[the remark this answers is not in the document]
Comment (asutula) [reply cmd/rigs/cmd/publish_data.go:209]: Issue created #237

File: pkg/builder/builder.go [hunk 5]
```diff
@@ -34,7 +32,8 @@ const (
 
 var (
 	defaultRenderConfig = renderConfig{
-		size:        1200,
+		size:        4000,
+		mediumSize:  2000,
```
Comment (asutula) [inline pkg/builder/builder.go:36]: Lots of lines changed in here, but it just adds support for rendering a medium sized image and, importantly, cleans up the rendering pipeline to be way easier to understand and debug. It works, don't worry about the details.

File: pkg/storage/tableland/common/sqlfactory.go [hunk 6]
```diff
@@ -38,45 +39,23 @@ func (s *SQLFactory) SQLForInsertingParts(table string, parts []local.Part) (str
 }
 
 // SQLForInsertingLayers returns the SQL statement.
-func (s *SQLFactory) SQLForInsertingLayers(table string, cid string, layers []local.Layer) (string, error) {
+func (s *SQLFactory) SQLForInsertingLayers(table string, layers []local.Layer) (string, error) {
```
Comment (asutula) [inline pkg/storage/tableland/common/sqlfactory.go:42]: Changes in here just to account for the fact that imagery root cids are now stored in a lookups table.

File: pkg/storage/tableland/impl/files/store.go [hunk 7]
```diff
@@ -64,62 +64,67 @@ func (s *Store) InsertParts(ctx context.Context, parts []local.Part) error {
 }
 
 // InsertLayers implements InsertLayers.
-func (s *Store) InsertLayers(ctx context.Context, cid string, layers []local.Layer) error {
+func (s *Store) InsertLayers(ctx context.Context, layers []local.Layer) error {
```
Comment (asutula) [inline pkg/storage/tableland/impl/files/store.go:67]: Same thing, imagery cid now stored in the lookups table, so not needed in this file anymore.

File: pkg/storage/tableland/impl/sqlite/store.go [hunk 8]
```diff
@@ -54,8 +54,8 @@ func (s *Store) InsertParts(ctx context.Context, parts []local.Part) error {
 }
 
 // InsertLayers implements InsertLayers.
-func (s *Store) InsertLayers(ctx context.Context, cid string, layers []local.Layer) error {
-	sql, err := s.factory.SQLForInsertingLayers(tableland.LayersDefinition.Prefix, cid, layers)
+func (s *Store) InsertLayers(ctx context.Context, layers []local.Layer) error {
+	sql, err := s.factory.SQLForInsertingLayers(tableland.LayersDefinition.Prefix, layers)
```
Comment (asutula) [inline pkg/storage/tableland/impl/sqlite/store.go:58]: Same thing, imagery cid now stored in the lookups table, so not needed in this file anymore.

Comment (asutula) [reply pkg/storage/tableland/impl/sqlite/store.go:58]: Plus some other cleanup.

File: pkg/storage/tableland/impl/tableland/store.go [hunk 9]
```diff
@@ -73,92 +73,105 @@ func (s *Store) InsertParts(ctx context.Context, parts []local.Part) error {
 }
 
 // InsertLayers implements InsertLayers.
-func (s *Store) InsertLayers(ctx context.Context, cid string, layers []local.Layer) error {
+func (s *Store) InsertLayers(ctx context.Context, layers []local.Layer) error {
```
Comment (asutula) [inline pkg/storage/tableland/impl/tableland/store.go:76]: Same thing, imagery cid now stored in the lookups table, so not needed in this file anymore. Plus some cleanup.

File: pkg/storage/tableland/store.go [hunk 10]
```diff
@@ -33,22 +33,10 @@ var (
 			fleet text not null,
 			rig_attributes_value text not null,
 			position integer not null,
-			cid text not null,
+			path text not null,
```
Comment (asutula) [inline pkg/storage/tableland/store.go:36]: Since we're moving to the new schema design, this was changed to a more appropriate column name (we don't store any cid in there, just the path within a root cid).

File: pkg/storage/tableland/store.go [hunk 11]
```diff
@@ -60,17 +48,58 @@ var (
...
+		Prefix: "pilot_sessions",
+		Schema: `(
+			id integer primary key,
+  		rig_id integer not null,
+  		pilot_contract text,
+  		pilot_id integer,
+  		start integer not null,
+  		end integer
+		)`,
+	}
```
Comment (asutula) [inline pkg/storage/tableland/store.go:77]: New `pilot_sessions` table. This is basically the same schema I sent @datadanne with the slight change that `start_time` and `end_time` are now `start` and `end`. I haven't heard that there is anything lacking with this schema, so I think we're good to go here, but let me know if anything should change @datadanne.

File: cmd/rigs/cmd/publish_data.go [hunk 12]
```diff
@@ -218,3 +180,35 @@ func attrsJobs(ctx context.Context, s local.Store, jobID *int) ([]wpool.Job, err
...
+			lookups := tableland.Lookups{
+				RendersCid:           rendersCid,
+				LayersCid:            layersCid,
+				ImageFullName:        "image_full.png",
+				ImageFullAlphaName:   "image_full_alpha.png",
+				ImageMediumName:      "image_medium.png",
+				ImageMediumAlphaName: "image_medium_alpha.png",
+				ImageThumbName:       "image_thumb.png",
+				ImageThumbAlphaName:  "image_thumb_alpha.png",
+				AnimationBaseURL:     "https://tableland.xyz/rigs/animate?id=", // TODO: Update when we know it.
```
Comment (asutula) [inline cmd/rigs/cmd/publish_data.go:205]: This needs to be updated once we know the url of the animation app cc @joewagner

File: pkg/storage/tableland/store.go (hunk 11 above)
Comment (datadanne) [reply pkg/storage/tableland/store.go:77]: Me and dtb have been discussing the schema a bit here: https://github.com/tablelandnetwork/rigs/commit/1d97b32ba315758d3ddc2a8a22a9bf4ece9bbf1b#r86016113

tldr: I think we should add a `owner text` column so that we can record who the rig owner was for every pilot sessions

Comment (asutula) [reply pkg/storage/tableland/store.go:77]: Cool, added it. Also switched back to `start_time` and `end_time` to avoid using the reserved word `end`.

**Comment by asutula** [comment]
Thanks for that perspective @datadanne. We should definitely document that. I'm planning on a new blog post explaining these design changes and will mention it there as well. Just to clarify, if you know that every minted token will have at least one attribute, can you just rely on the attributes table to do your "is it minted?" check?

File: cmd/rigs/cmd/local_render.go [hunk 13]
```diff
@@ -18,7 +18,8 @@ func init() {
 
 	renderCmd.Flags().String("to-path", "./renders", "path to write the images to")
 	renderCmd.Flags().String("layers-path", "./artifacts/layers", "path to the rigs layer images")
-	renderCmd.Flags().Int("size", 1200, "width and height of generated images")
+	renderCmd.Flags().Int("size", 4000, "width and height of generated images")
+	renderCmd.Flags().Int("medium-size", 2000, "width and height of generated medium images")
```
[the remark this answers is not in the document]
Comment (sanderpick) [reply cmd/rigs/cmd/local_render.go:22]: nice, sgtm

**Review by sanderpick — APPROVED** [review]
I didn't dive into understanding all the implemention, but conceptually all looks good to me. nice work!
